homeassistant/components/nobo_hub/sensor.py

```diff
@@ -0,0 +1,100 @@
+"""Python Control of Nobø Hub - Nobø Energy Control."""
+from __future__ import annotations
+
+from pynobo import nobo
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    ATTR_IDENTIFIERS,
+    ATTR_MANUFACTURER,
+    ATTR_MODEL,
+    ATTR_NAME,
+    ATTR_SUGGESTED_AREA,
+    ATTR_VIA_DEVICE,
+    TEMP_CELSIUS,
+)
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
+
+from .const import ATTR_SERIAL, ATTR_ZONE_ID, DOMAIN, NOBO_MANUFACTURER
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up any temperature sensors connected to the Nobø Ecohub."""
+
+    # Setup connection with hub
+    hub: nobo = hass.data[DOMAIN][config_entry.entry_id]
+
+    async_add_entities(
+        [
+            NoboTemperatureSensor(component["serial"], hub)
+            for component in filter(
+                lambda component: component[ATTR_MODEL].has_temp_sensor,
+                hub.components.values(),
+            )
+        ],
+        True,
```

Comment:
It was done in this way in the `climate` platform to make sure that the callback is registered in `async_added_to_hass` before reading the initial state. We may lose an update if we first read the state, and then register the callback.

Reply:
The entity update during entity addition is done before `async_added_to_hass` is called.

Reply:
I don't follow the logic in this answer regardless of the order of operations, as we call read state in the callback too. The entity state is always updated at the end of entity addition so we don't need the callback for that.

Reply:
Thinking a bit more about this, to avoid the race condition I moved the call to `_read_state()` to `async_added_to_hass()` after the callback is registered. Note that `_read_state()` only does local calls.

The climate platform should probably have a similar change.

Reply:
What is causing the race? What side effects does `get_current_component_temperature` have?

Reply:
`get_current_component_temperature` doesn't have any side effects, it only gets the in-memory state from the `self._nobo` object, but there is a task in the `pynobo` library reading updates from the TCP socket that is already initialised.

I am worried about the case for the task receiving a message before the callback is set up, i.e., this order of calls:
1. `get_current_component_temperature()`
2. [`socket_receive()`](https://github.com/echoromeo/pynobo/blob/master/pynobo.py#L669)
   a. Updates internal state ([`response_handler()`](https://github.com/echoromeo/pynobo/blob/master/pynobo.py#L700))  
   b. Calls [callbacks](https://github.com/echoromeo/pynobo/blob/master/pynobo.py#L681-L682)
4. `register_callback()`

In this case, the state in the sensor loose an update.

I realise that in Python it doesn't matter if the call order is 
```
self._nobo.register_callback(self._after_update)
self._read_state()
```
or
```
self._read_state()
self._nobo.register_callback(self._after_update)
```
as long as `register_callback()` is not async, as the task would not get a chance to run. But is it guaranteed that the `socket_receive` task doesn't run between `__init_()` and `async_added_to_hass()`?

Reply:
What starts the background task? If we want to wait until all platforms are ready, we should not start the background task until they are ready.

Reply:
I see. The background task is currently started in `async_setup_entry`:
https://github.com/oyvindwe/home-assistant/blob/nobo_hub-sensor/homeassistant/components/nobo_hub/__init__.py#L36

I can move it to after the call to `async_forward_entry_setups`

Reply:
Good!

Reply:
Hmm - I need create a PR to refactor the `pynobo` library so it is possible to do the initial connect and retrieve the data required to create the devices and entities without starting the background task (all this information is provided on the same socket as status updates).

Reply:
`pynobo` released and this PR is now updated as suggested above.